File: JavaScript/4-Stack/Algorithms/2-NextGreaterElement/NextGreaterElementPro.js
/*
    Next Greater Element Problem:

    Given an array, find the next greater element (NGE) for each element. The next greater element for an element x is the first element to the right of x that is greater than x. If no such element exists, the NGE is -1.

    Here are five different approaches to solve this problem, along with real-world project implementation examples:
*/

// 1. Brute Force Approach
//    - Iterate through the array.
//    - For each element, iterate through the rest of the array to find the first greater element.
//
//    - Real-world example:  Analyzing website traffic logs.  For each page visit, find the next page visited by the same user that had a higher bounce rate.
//
function nextGreaterElement_BruteForce(arr) {
    const n = arr.length;
    const result = new Array(n).fill(-1);

    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
            if (arr[j] > arr[i]) {
                result[i] = arr[j];
                break;
            }
        }
    }
    return result;
}

// 2. Stack Approach
//    - Use a stack to keep track of elements.
//    - Iterate through the array from right to left.
//    - For each element:
//        - Pop elements from the stack that are less than or equal to the current element.
//        - The next greater element is the top of the stack (or -1 if the stack is empty).
//        - Push the current element onto the stack.
//
//    - Real-world example:  Stock price analysis.  For each day's stock price, find the next day's price that is higher.
//
function nextGreaterElement_Stack(arr) {
    const n = arr.length;
    const result = new Array(n).fill(-1);
    const stack = [];

    for (let i = n - 1; i >= 0; i--) {
        while (stack.length > 0 && stack[stack.length - 1] <= arr[i]) {
            stack.pop();
        }
        result[i] = stack.length > 0 ? stack[stack.length - 1] : -1;
        stack.push(arr[i]);
    }
    return result;
}

// 3. Optimized Stack Approach (Iterate Once)
//   - Uses a stack, but iterates through the array only once.
//   - Iterate through the array.
//   - For each element:
//     - While the stack is not empty and the current element is greater than the element at the top of the stack:
//       - Pop the element from the stack.
//       - The current element is the NGE for the popped element.
//     - Push the current element's index onto the stack.
//   - After the loop, any elements remaining in the stack do not have an NGE in the array.
//
//   - Real-world example:  Finding trends in sensor data.  For each sensor reading, find the next reading that exceeds a threshold.
function nextGreaterElement_OptimizedStack(arr) {
    const n = arr.length;
    const result = new Array(n).fill(-1);
    const stack = [];

    for (let i = 0; i < n; i++) {
        while (stack.length > 0 && arr[i] > arr[stack[stack.length - 1]]) {
            const top = stack.pop();
            result[top] = arr[i];
        }
        stack.push(i);
    }
    return result;
}

// 4. Using a Map with Stack
//    - Use a stack to keep track of elements' indices.
//    - Use a Map to store the NGE for each element.
//    - Iterate through the array.
//    - For each element:
//        - Pop elements from the stack whose values are less than the current element.
//        - Store the current element as the NGE for the popped elements in the Map.
//        - Push the current element's index onto the stack.
//    - Iterate through the array and use the Map to populate the result array.
//
//    - Real-world example:  Compiler design.  For each variable declaration, find the next declaration of a variable with a larger scope.
//
function nextGreaterElement_MapAndStack(arr) {
    const n = arr.length;
    const result = new Array(n).fill(-1);
    const stack = [];
    const ngeMap = new Map();

    for (let i = 0; i < n; i++) {
        while (stack.length > 0 && arr[i] > arr[stack[stack.length - 1]]) {
            const top = stack.pop();
            ngeMap.set(top, arr[i]);
        }
        stack.push(i);
    }

    for (let i = 0; i < n; i++) {
        result[i] = ngeMap.get(i) || -1;
    }
    return result;
}

// 5. Circular Array Approach
//    - To find the next greater element in a circular array (where the last element's next is the first element),
//      we can effectively double the array by iterating twice.
//    - Use a stack to keep track of indices.
//    - Iterate through the doubled array (or iterate twice through the original).
//    - For each element:
//        - Pop elements from the stack whose values are less than the current element.
//        - Update the NGE for the popped elements.
//        - Push the current element's index onto the stack.  Use the modulo operator (%) to handle the circular indexing.
//
//    - Real-world example:  Analyzing server request patterns in a loop.  For each request, find the next request in the cycle that has a higher priority.
//
function nextGreaterElement_CircularArray(arr) {
    const n = arr.length;
    const result = new Array(n).fill(-1);
    const stack = [];

    for (let i = 0; i < 2 * n; i++) {
        while (stack.length > 0 && arr[i % n] > arr[stack[stack.length - 1]]) {
            const top = stack.pop();
            result[top] = arr[i % n];
        }
        stack.push(i % n);
    }
    return result;
}

// Example Usage and Output
const array = [4, 5, 2, 25, 7, 8, 1];
console.log("Original Array:", array);

console.log("\n1. Brute Force Approach:", nextGreaterElement_BruteForce(array));
console.log("   Expected Output:    [ 5, 25, 25, -1, 8, -1, -1 ]");

console.log("\n2. Stack Approach:", nextGreaterElement_Stack(array));
console.log("   Expected Output:    [ 5, 25, 25, -1, 8, -1, -1 ]");

console.log("\n3. Optimized Stack Approach:", nextGreaterElement_OptimizedStack(array));
console.log("   Expected Output:    [ 5, 25, 25, -1, 8, -1, -1 ]");

console.log("\n4. Map and Stack Approach:", nextGreaterElement_MapAndStack(array));
console.log("   Expected Output:    [ 5, 25, 25, -1, 8, -1, -1 ]");

const circularArray = [1, 2, 1];
console.log("\nOriginal Circular Array:", circularArray);
console.log("5. Circular Array Approach:", nextGreaterElement_CircularArray(circularArray));
console.log("   Expected Output:    [ 2, -1, 2 ]");

File: Java/4-Stack/Algorithms/2-NextGreaterElement/NextGreaterElementPro.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public class NextGreaterElement {

    public static void main(String[] args) {
        int[] arr = {4, 6, 3, 2, 8, 1};
        int[] nextGreaterElement;

        // Approach 1: Brute Force
        nextGreaterElement = nextGreaterElementBruteForce(arr);
        System.out.println("Next Greater Element (Brute Force): " + Arrays.toString(nextGreaterElement)); // Output: [6, 8, 8, 8, -1, -1]

        // Approach 2: Stack
        nextGreaterElement = nextGreaterElementStack(arr);
        System.out.println("Next Greater Element (Stack): " + Arrays.toString(nextGreaterElement)); // Output: [6, 8, 8, 8, -1, -1]

        // Approach 3: Optimized Stack (Handles Duplicates)
        nextGreaterElement = nextGreaterElementOptimizedStack(arr);
        System.out.println("Next Greater Element (Optimized Stack): " + Arrays.toString(nextGreaterElement));

        // Approach 4: Using HashMap for mapping (for finding NGE of specific elements)
        int[] resultWithMap = nextGreaterElementWithMap(arr);
        System.out.println("Next Greater Element (HashMap): " + Arrays.toString(resultWithMap));

        // Approach 5: Circular Array
        int[] circularArray = {1, 5, 3, 6, 4};
        nextGreaterElement = nextGreaterElementCircular(circularArray);
        System.out.println("Next Greater Element (Circular Array): " + Arrays.toString(nextGreaterElement)); //Output: [5, 6, 6, -1, 5]
    }

    /**
     * Approach 1: Brute Force
     * -   Time Complexity: O(n^2)
     * -   Space Complexity: O(n)
     * -   Real-world application: Simple data analysis where you need to compare each data point with every other data point.  For example, finding the next higher stock price for each day in a short time frame.
     *
     * @param arr The input array.
     * @return An array where each element is the next greater element in the input array.
     */
    public static int[] nextGreaterElementBruteForce(int[] arr) {
        int n = arr.length;
        int[] result = new int[n];
        for (int i = 0; i < n; i++) {
            result[i] = -1; // Default value if no greater element is found
            for (int j = i + 1; j < n; j++) {
                if (arr[j] > arr[i]) {
                    result[i] = arr[j];
                    break;
                }
            }
        }
        return result;
    }

    /**
     * Approach 2: Stack
     * -   Time Complexity: O(n)
     * -   Space Complexity: O(n)
     * -   Real-world application: Processing real-time data streams, such as stock prices or sensor readings, where you need to efficiently track the next significant event.  For example, in a stock market analysis, identify the next peak price after each day.
     *
     * @param arr The input array.
     * @return An array where each element is the next greater element in the input array.
     */
    public static int[] nextGreaterElementStack(int[] arr) {
        int n = arr.length;
        int[] result = new int[n];
        Arrays.fill(result, -1); // Initialize result array with -1
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && arr[i] > arr[stack.peek()]) {
                result[stack.pop()] = arr[i];
            }
            stack.push(i);
        }
        return result;
    }

    /**
     * Approach 3: Optimized Stack (Handles Duplicates)
     * -   Time Complexity: O(n)
     * -   Space Complexity: O(n)
     * -   Real-world application: Analyzing data with duplicate values, such as website traffic logs or sales data, where you need to find the next greater event, even if duplicates exist.  For example, finding the next higher sales amount for each transaction, even if some transactions have the same amount.
     *
     * @param arr The input array.
     * @return An array where each element is the next greater element in the input array.
     */
    public static int[] nextGreaterElementOptimizedStack(int[] arr) {
        int n = arr.length;
        int[] result = new int[n];
        Arrays.fill(result, -1);
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && arr[i] > arr[stack.peek()]) {
                result[stack.pop()] = arr[i];
            }
            // Important: Push the *index*, not the value.  This allows us to handle duplicates correctly.
            stack.push(i);
        }
        return result;
    }

    /**
     * Approach 4: Using HashMap for mapping (for finding NGE of specific elements)
     * -   Time Complexity: O(n)
     * -   Space Complexity: O(n)
     * -   Real-world application:  Useful when you need to quickly look up the next greater element for specific values, rather than for every element in the array.  For example, in a system that monitors stock prices, you might want to quickly find the next higher price for a specific set of target stocks.
     *
     * @param arr The input array.
     * @return A Map where key is the element and value is the next greater element.
     */
    public static int[] nextGreaterElementWithMap(int[] arr) {
        int n = arr.length;
        int[] result = new int[n];
        Arrays.fill(result, -1);
        Stack<Integer> stack = new Stack<>();
        Map<Integer, Integer> ngeMap = new HashMap<>(); // Map to store element and its NGE

        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && arr[i] > arr[stack.peek()]) {
                int poppedIndex = stack.pop();
                ngeMap.put(arr[poppedIndex], arr[i]); // Store the mapping
                result[poppedIndex] = arr[i];
            }
            stack.push(i);
        }

        return result;
    }

    /**
     * Approach 5: Circular Array
     * -   Time Complexity: O(n)
     * -   Space Complexity: O(n)
     * -   Real-world application: Analyzing cyclic data, such as server activity logs or network traffic patterns, where the last element is considered to be adjacent to the first element. For example, finding the next peak server load in a 24-hour cycle.
     *
     * @param arr The input circular array.
     * @return An array where each element is the next greater element in the circular input array.
     */
    public static int[] nextGreaterElementCircular(int[] arr) {
        int n = arr.length;
        int[] result = new int[n];
        Arrays.fill(result, -1);
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < 2 * n; i++) { // Iterate twice to simulate circular array
            int index = i % n; // Get the actual index in the array
            while (!stack.isEmpty() && arr[index] > arr[stack.peek()]) {
                result[stack.pop()] = arr[index];
            }
            if (i < n) { // Push only the first 'n' indices
                stack.push(index);
            }
        }
        return result;
    }
}

